dali/pipeline/operator/sequence_operator.h

```diff
@@ -0,0 +1,607 @@
+      const auto &expand_desc = GetOutputExpandDesc(ws, output_idx);
+      const auto &shape = output_desc[output_idx].shape;
+      DALI_ENFORCE(shape.num_samples() == expand_desc.NumExpanded(),
+                   make_string("Unexpected number of frames inferred in the operator for output ",
+                               output_idx, ". Expected ", expand_desc.NumExpanded(),
+                               " but the operator returned ", shape.num_samples(), "."));
+      output_desc[output_idx].shape = fold_outermost_like(shape, expand_desc.DimsToExpand());
+    }
+  }
+
+  const ExpandDesc &GetInputExpandDesc(int input_idx) const {
+    DALI_ENFORCE(0 <= input_idx && static_cast<size_t>(input_idx) < input_expand_desc_.size());
+    return input_expand_desc_[input_idx];
+  }
+
+  void SetupSequenceOperator(const workspace_t<Backend> &ws) {
+    auto num_inputs = ws.NumInput();
+    input_expand_desc_.clear();
+    input_expand_desc_.reserve(num_inputs);
+    for (int input_idx = 0; input_idx < num_inputs; input_idx++) {
+      const auto &input_shape = ws.GetInputShape(input_idx);
+      const auto &layout = GetInputLayout(ws, input_idx);
+      input_expand_desc_.emplace_back(input_shape, layout, ShouldExpandChannels(input_idx));
+    }
+    expand_like_idx_ = InferExpandableInputIdx(ws);
+  }
+
+  int InferExpandableInputIdx(const workspace_t<Backend> &ws) {
+    auto num_inputs = ws.NumInput();
+    for (int input_idx = 0; input_idx < num_inputs; input_idx++) {
+      if (input_expand_desc_[input_idx].NumDimsToExpand() > 0) {
+        return input_idx;
+      }
+    }
+    return -1;
+  }
+
+  /**
+   * @brief Returns the index of the operator input which is expandable and other
+   * inputs should be expanded or broadcasted in a manner consistent with the
+   * given input. If no input is expandable, returns -1.
+   */
+  int GetReferentialInputIdx() const {
+    return expand_like_idx_;
+  }
+
+  void ExpandInputs(const workspace_t<Backend> &ws) {
+    for (int input_idx = 0; input_idx < ws.NumInput(); input_idx++) {
+      ExpandInput(ws, input_idx);
+    }
+  }
+
+  void ExpandOutputs(const workspace_t<Backend> &ws) {
+    for (int output_idx = 0; output_idx < ws.NumOutput(); output_idx++) {
+      ExpandOutput(ws, output_idx);
+    }
+  }
+
+  void ExpandArguments(const ArgumentWorkspace &ws) {
+    for (const auto &arg_input : ws) {
+      auto &shared_tvec = arg_input.second.tvec;
+      assert(shared_tvec);
+      ExpandArgument(ws, arg_input.first, *shared_tvec);
+    }
+  }
+
+  bool HasPerFrameArgInputs(const workspace_t<Backend> &ws) {
+    for (const auto &arg_input : ws) {
+      auto &shared_tvec = arg_input.second.tvec;
+      assert(shared_tvec);
+      if (detail::is_per_frame(*shared_tvec)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  void VerifyExpanionConsistency(int expand_idx, const ExpandDesc &expand_desc, int input_idx,
```

Comment:
```suggestion
  void VerifyExpansionConsistency(int expand_idx, const ExpandDesc &expand_desc, int input_idx,
```
I know, it was my typo in the suggestion. Sorry for that.

Reply:
I could catch that on the other hand.
Done.